### PR TITLE
tensorflow distribute training on k8s

### PR DESCRIPTION
**Pull request by lwpyr**

### CODE REVIEW

**Review by terrytangyuan — COMMENTED** [review]
Why does this need to be in ElasticDL repo?

**Comment by skydoorkai** [comment]
Why each of  `master/client/allreduce` has its own Dockerfile? Their relationship is not clear.

**Comment by lwpyr** [comment]
This is a task assigned to me as an intern.
allreduce is an example training script from tensorflow official tutorial but modified to remove bugs and support auto deployment by this demo. 
the dockerfile inside the client is a mistake. actually it's a intermediate file created by the client

**Comment by QiJune** [comment]
@terrytangyuan We want to do some benchmark on ElasticDL, and compare it with the internal distribution strategy in TensorFlow. So what's your opinion, do we need another repo to put these benchmark scripts?

**Comment by terrytangyuan** [comment]
It should be fine in this repo as it’s in a separate benchmark folder.

File: benchmark/tensorflow-runtime/allreduce/exec.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
+# allreduce training example
```
Comment (terrytangyuan) [inline benchmark/tensorflow-runtime/allreduce/exec.py]: Remove unnecessary comments

File: benchmark/tensorflow-runtime/allreduce/exec.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+from __future__ import absolute_import, division, print_function, unicode_literals
```
Comment (terrytangyuan) [inline benchmark/tensorflow-runtime/allreduce/exec.py]: Why is this needed? We are only using Python 3

File: benchmark/tensorflow-runtime/allreduce/exec.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+import os
+
+strategy = tf.distribute.experimental.MultiWorkerMirroredStrategy()
+tfds.disable_progress_bar()
+tf.compat.v1.disable_eager_execution()
+
+BUFFER_SIZE = 10000
+BATCH_SIZE = 64
+
+# 将 MNIST 数据从 (0, 255] 缩放到 (0., 1.]
```
Comment (terrytangyuan) [inline benchmark/tensorflow-runtime/allreduce/exec.py]: Use English

File: benchmark/tensorflow-runtime/allreduce/exec.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+      tf.keras.layers.Dense(10, activation='softmax')
+  ])
+  model.compile(
+      loss=tf.keras.losses.sparse_categorical_crossentropy,
+      optimizer=tf.keras.optimizers.SGD(learning_rate=0.001),
+      metrics=['accuracy'])
+  return model
+
+NUM_WORKERS = int(os.environ["NUM_WORKERS"])
+print(NUM_WORKERS)
```
Comment (terrytangyuan) [inline benchmark/tensorflow-runtime/allreduce/exec.py]: Unnecessary print

File: benchmark/tensorflow-runtime/allreduce/exec.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+GLOBAL_BATCH_SIZE = 64 * NUM_WORKERS
+train_datasets = train_datasets_unbatched.batch(GLOBAL_BATCH_SIZE)
+with strategy.scope():
+  multi_worker_model = build_and_compile_cnn_model()
+multi_worker_model.fit(x=train_datasets, epochs=10, steps_per_epoch=60000/GLOBAL_BATCH_SIZE) # 60000 is the size of the training set
```
Comment (terrytangyuan) [inline benchmark/tensorflow-runtime/allreduce/exec.py]: Move this number to a constant and then the comment is not necessary

File: benchmark/tensorflow-runtime/client/client.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+args = parser.parse_args()
+
+build_pod_image(args.image, args.image+'_pod')
+params = {}
+params["strategy"] = args.strategy
+params["n_worker"] = args.worker_num
+params["n_ps"] = args.ps_num
+params["img"] = args.image+'_pod'
+params["name"] = args.task
+print(args.master)
```
Comment (terrytangyuan) [inline benchmark/tensorflow-runtime/client/client.py:33]: Unnecessary print

File: benchmark/tensorflow-runtime/client/cmd [hunk 7]
```diff
@@ -0,0 +1,6 @@
+python client.py --master "http://192.168.64.5:31898" \
```
Comment (terrytangyuan) [inline benchmark/tensorflow-runtime/client/cmd]: Is this a private IP address?

File: benchmark/tensorflow-runtime/readme.MD [hunk 8]
```diff
@@ -0,0 +1,37 @@
+This is a demo of tensorflow distribute training on K8s
```
Comment (terrytangyuan) [inline benchmark/tensorflow-runtime/readme.MD:1]: TensorFlow

Comment (terrytangyuan) [inline benchmark/tensorflow-runtime/readme.MD:1]: File name MD -> md

File: benchmark/tensorflow-runtime/allreduce_keras_example/exec.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+  return image, label
+
+datasets, info = tfds.load(name='mnist',
+                           with_info=True,
+                           as_supervised=True)
+
+train_datasets_unbatched = datasets['train'].map(scale).cache().shuffle(BUFFER_SIZE).repeat()
+
+def build_and_compile_cnn_model():
+  model = tf.keras.Sequential([
```
Comment (skydoorkai) [inline benchmark/tensorflow-runtime/allreduce_keras_example/exec.py:32]: Since we are using the MNIST dataset, better to use the same Keras model in model_zoo/mnist_functional_api/ for comparison.

**Comment by skydoorkai** [comment]
Closing. We may revisit the benchmark later.
